### DISCUSSION TITLE
What happened to Country and Language tags?

### DISCUSSION THREAD

**Discussion by dev-onetv** [iptv-org/community#53] [Q&A] [answered]
Hi there, the country and language tags were extremely helpful and many people use the main index.m3u file that without the tags is not possible.

I am very disappointed that those are removed. Would it be possible to restore the tags?

Thank you very much for all your efforts.

**Comment by BellezaEmporium** [comment]
If you need this, use the Country or Language playlist. It is the same as the main index.m3u file.

**Reply by dev-onetv** [reply]
Thank you very much for your reply. That is certainly possible.

But having the tags in the main index.m3u would have the advantage of narrowing down search by language and country without the need for combining all those countries and languages data.

I think it would be extremely useful if you could keep the country and language tags in the main index.m3u file as before.

Thanks again

**Reply by BellezaEmporium** [reply]
> Thank you very much for your reply. That is certainly possible.
> 
> But having the tags in the main index.m3u would have the advantage of narrowing down search by language and country without the need for combining all those countries and languages data.
> 
> I think it would be extremely useful if you could keep the country and language tags in the main index.m3u file as before.
> 
> Thanks again

I'm pretty sure that if you manage to get both files, you can merge the EXTINF sectors to have both language and country using a script. We discarded those due to lack of usage.

**Reply by dev-onetv** [reply]
Thank you @LaneSh4d0w. For sure it is possible to merge those. However, I was thinking if it is not causing any harm, please keep the tags. 

The tags are certainly useful, IP TV players like ours can conveniently search channels based on these tags without requiring to download and merge multiple files on the client side.

In fact wihout those tags having a list of more than 7000 channels in one playlist may not be useful.

**Comment by freearhey** [comment, marked as answer]
The `tvg-country` and `tvg-language` attributes were removed from playlists in #7809. This was done, as @LaneSh4d0w correctly pointed out, because most players simply do not use these attributes.

Also if before there really was no other way to filter channels by language or country, now you can get a list of all channels through our [API](https://github.com/iptv-org/api#channels) and filter/sort them however you want.

**Reply by dev-onetv** [reply]
@freearhey: Thank for getting back to me and your suggestion. Yes, definitely using the new APIs would be a better solution. Thank you very much for all your efforts. 

May I suggest that if it is possible to have another API that has all attributes for a channel (i.e., combining channels, streams, guides, ...).

This can improve the performance on IP TV players by reducing number of URL requests to only one and not requiring combining the json files.

Thanks again for your answers.

**Reply by BellezaEmporium** [reply]
The API link already contains all the channels, simply sort the channels out by what you need in terms of language and country.

**Reply by dev-onetv** [reply]
@LaneSh4d0w: Thank you for your reply. I understand it is possible to process all the files and combine them. 

I understand you and  @freearhey already responded to my request for adding back `tvg-country`, and `tvg-language` to m3u files. While we can definitely process the m3u files and combine them, many users would prefer to use their own m3u list which they often take them from this repository.

Having those attributes in m3u file is extremely useful to search and narrowing down the channel list.

I understand you mentioned many of main stream m3u players do not use those attributes. But you may not remove those attributes, which are part of m3u standard, just based on how main stream m3u players currently function. This will prevent creating new and better solutions.

We would be extremely grateful, if you could kindly reconsider adding back the`tvg-country`, and `tvg-language` to m3u files.

Thank you very much again for all your efforts.

**Reply by freearhey** [reply]
First, there is no m3u standard and, as far as I know, most third-party playlists did not have these attributes.

Second, you don't need to change anything for your users. 

Every link in our playlists has a `tvg-id` attribute with a unique channel ID (ex: `CNN.us`), you just need to load the [api/channels.json](https://github.com/iptv-org/api#channels) into your app, find the channel data by `id` and use it to filtering instead.

Third, as for combining `channels.json`, `streams.json`, and `guides.json` into one file. This was not done precisely because of their size. It's much faster to load 3 files in parallel than one huge file (`channels.json` alone already exceeds 10Mb) and since GitHub support [HTTP/2](https://web.dev/performance-http2/#request-and-response-multiplexing) protocol, sending parallel requests doesn't affect performance in any way.

But if GitHub adds support for dynamic content to Pages in the future, and there's no longer a need to download whole files, we'll definitely revisit this issue.

I hope this helps!

**Reply by dev-onetv** [reply]
@freearhey: Thank for getting back to me on this. You are correct M3U is an extension of a stranded.

Now it is clear to me why you decided not to combine `channels.json`, `streams.json`, and `guides.json`. Thanks for clarifying this.

As for M3U playlist, yes it is certainly possible to use `channels.json` as an additional database. But there are two main problems with this solution: First as you pointed out the `channels.json` exceeds 10 MB. Second it will not be clean solution and the solution will be tied to this repository (This is not scalable solution).

Our argument for suggesting to add back `tv-language` and `tvg-country` are:

1. Language and country are most important information for a channel in a playlist (as important as the logo and channel name and group title).
2. Adding attributes to #EXTINF is a perfect place to add language and country information (as you alreardy have many useful attributes there).
3. This will be a clean, simple and scalable solution that can be adapted by other players and possibly future standards.
4. Adding these flags has a minor impact on increasing the files volume (the main index.m3u file increases to ~2 MB from ~1.7 MB).
5. Thanks for the great structure of database and code base that you have developed in this repository, adding back thee attributes will not take too much time (based on what we can see in #7809).

**Reply by freearhey** [reply]
> First as you pointed out the channels.json exceeds 10 MB

Why not just include `channels.json` in your build then.

> Second it will not be clean solution and the solution will be tied to this repository (This is not scalable solution).

Well if you plan to use playlists from other sources then using unique attributes from our playlists is definitely not the best solution.

**Reply by dev-onetv** [reply]
> Why not just include `channels.json` in your build then.

This is certainly possible or it can be cached, however this would be inefficient solution as it adds 10 MB to APK size, need further processing and keeping separate database. Also, since playlist keeps being updates, the `channels.json` should be updated/cached again frequently too.
 
> Well if you plan to use playlists from other sources then using unique attributes from our playlists is definitely not the best solution.

This repository is certainly the most well known by users, most accessible to everyone, and probably most used in the world. The other providers may (and should) follow and provide attributes for language and country (especially if the playlist is large). 

Our request is not solely for our app benefit. We believe having the language and country attributes are useful and other M3U players can implement functionality to improve user experience. Without having functionality for search by languages and country, no user can go through these large number of channels in one playlist to find what they want to watch. Also there are many countries with same language and some countries with several languages. User usually want to narrow down channels in playlists by both country and language.

Lastly, having these attributes may help in collecting the language and country data when someone contributes to this repository (as users contribute to country playlist file, I am not sure how language data is collected after someone contributes to a country playlist file).
